### PR TITLE
Name for custom newsletters

### PR DESCRIPTION
**Pull request by vershwal**
refs https://github.com/TryGhost/Product/issues/3862 & https://github.com/TryGhost/Product/issues/3863

### CODE REVIEW

File: ghost/members-api/lib/controllers/RouterController.js [hunk 1]
```diff
@@ -483,19 +485,46 @@ module.exports = class RouterController {
...
-                        const invalidNewsletterIds = newsletterIds.filter(id => !validNewsletterIds.includes(id));
+                    if (newsletters.length !== newsletterNames.length) {
+                        if (newsletters.length < newsletterNames.length) { //check for invalid newsletters
+                            const validNewsletters = newsletters.map(newsletter => `'${newsletter.name}'`);
+                            const invalidNewsletters = newsletterNames.filter(newsletter => !validNewsletters.includes(newsletter));
+
+                            throw new errors.BadRequestError({
+                                message: tpl(messages.invalidNewsletters, {newsletters: invalidNewsletters})
+                            });
+                        } else { //check for ambiguous newsletters
```
Comment (vershwal) [inline ghost/members-api/lib/controllers/RouterController.js:503]: Need to confirm if this is even possible. While testing I found that we can't create two newsletters with one name. Even if a newsletter is archived, it doesn't allow to create with the same name.

Comment (mike182uk) [reply ghost/members-api/lib/controllers/RouterController.js:503]: aha thats great, we can drop the code needed for ambiguity then 🥳

**Comment by vershwal** [comment]
@mike182uk updated this PR.

File: ghost/members-api/lib/controllers/RouterController.js [hunk 2]
```diff
@@ -17,7 +17,9 @@ const messages = {
     memberNotFoundSignUp: 'No member exists with this e-mail address. Please sign up first.',
     invalidType: 'Invalid checkout type.',
     notConfigured: 'This site is not accepting payments at the moment.',
-    invalidNewsletterId: 'Cannot subscribe to invalid newsletter {ids}'
+    invalidNewsletters: 'Cannot subscribe to invalid newsletters {newsletters}',
+    archivedNewsletters: 'Cannot subscribe to archived newsletters {newsletters}',
+    ambiguousNewsletters: 'Cannot subscribe to ambiguous newsletters {newsletters}'
```
Comment (mike182uk) [inline ghost/members-api/lib/controllers/RouterController.js:22]: This is now redundant

File: ghost/members-api/lib/controllers/RouterController.js [hunk 3]
```diff
@@ -483,19 +485,30 @@ module.exports = class RouterController {
                 // Validate requested newsletters
                 let {newsletters: requestedNewsletters} = req.body;
 
-                if (requestedNewsletters && requestedNewsletters.length > 0) {
-                    const newsletterIds = requestedNewsletters.map(newsletter => newsletter.id);
+                if (requestedNewsletters && requestedNewsletters.length > 0 && requestedNewsletters.every(newsletter => newsletter.name !== undefined)) {
+                    const newsletterNames = requestedNewsletters.map(newsletter => newsletter.name);
+                    const newsletterNamesFilter = newsletterNames.map(newsletter => `'${newsletter.replace(/"/g, '\\"').replace(/'/g, `\\'`)}'`);
```
Comment (mike182uk) [inline ghost/members-api/lib/controllers/RouterController.js:490]: Could the `replace` here be simplified to using the result of a capture group?

```
.replace(/("|')/g, "\\$1")
```

File: ghost/members-api/test/unit/lib/controllers/router.test.js [hunk 4]
```diff
@@ -126,29 +126,33 @@ describe('RouterController', function () {
...
                         status: 'active'
                     }
                 ];
 
-                req.body.newsletters = newsletters.map(newsletter => ({id: newsletter.id}));
+                req.body.newsletters = newsletters.map(newsletter => ({name: newsletter.name}));
 
-                const newsletterIds = newsletters.map(newsletter => newsletter.id);
+                const newsletterNames = newsletters.map(newsletter => newsletter.name);
+                const newsletterNamesFilter = newsletterNames.map(newsletter => `'${newsletter.replace(/"/g, '\\"').replace(/'/g, `\\'`)}'`);
```
Comment (mike182uk) [inline ghost/members-api/test/unit/lib/controllers/router.test.js:147]: This can be updated to the more succinct version

**Review by mike182uk — APPROVED** [review]

**Pull request merged**